walk/git.go

```diff
@@ -1,169 +1,98 @@
+		// create a pipe to capture the command output
+		r, w := io.Pipe()
 
-			// we only want regular files, not directories or symlinks
-			if entry.Mode == filemode.Dir || entry.Mode == filemode.Symlink {
-				continue
-			}
-
-			// stat the file
-			path := filepath.Join(g.root, entry.Name)
+		// create a command which will execute from the specified sub path within root
+		cmd := exec.Command("git", "ls-files")
+		cmd.Dir = filepath.Join(g.root, g.path)
+		cmd.Stdout = w
 
-			info, err := os.Lstat(path)
-			if os.IsNotExist(err) {
-				// the underlying file might have been removed without the change being staged yet
-				g.log.Warnf("Path %s is in the index but appears to have been removed from the filesystem", path)
-				continue
-			} else if err != nil {
-				return fmt.Errorf("failed to stat %s: %w", path, err)
-			}
-
-			// determine a relative path
-			relPath, err := filepath.Rel(g.root, path)
-			if err != nil {
-				return fmt.Errorf("failed to determine a relative path for %s: %w", path, err)
-			}
-
-			file := File{
-				Path:    path,
-				RelPath: relPath,
-				Info:    info,
-			}
-
-			g.stats.Add(stats.Traversed, 1)
-			g.filesCh <- &file
-		}
-
-	default:
-
-		// read the git index into memory if it hasn't already
-		if idxCache == nil {
-			idxCache = &filetree{name: ""}
-			idxCache.readIndex(gitIndex)
-		}
-
-		// git index entries are relative to the repository root, so we need to determine a relative path for the
-		// one we are currently processing before checking if it exists within the git index
-		relPath, err := filepath.Rel(g.root, path)
-		if err != nil {
-			return fmt.Errorf("failed to find root relative path for %v: %w", path, err)
-		}
-
-		if !idxCache.hasPath(relPath) {
-			log.Debugf("path %s not found in git index, skipping", relPath)
-			return nil
-		}
-
-		err = filepath.Walk(path, func(path string, info fs.FileInfo, _ error) error {
-			// skip directories
-			if info.IsDir() {
-				return nil
-			}
-
-			// determine a path relative to g.root before checking presence in the git index
-			relPath, err := filepath.Rel(g.root, path)
-			if err != nil {
-				return fmt.Errorf("failed to determine a relative path for %s: %w", path, err)
-			}
-
-			if !idxCache.hasPath(relPath) {
-				log.Debugf("path %v not found in git index, skipping", relPath)
-				return nil
-			}
-
-			file := File{
-				Path:    path,
-				RelPath: relPath,
-				Info:    info,
-			}
-
-			g.stats.Add(stats.Traversed, 1)
-			g.filesCh <- &file
-			return nil
+		// execute the command in the background
+		g.eg.Go(func() error {
+			return w.CloseWithError(cmd.Run())
 		})
-		if err != nil {
-			return fmt.Errorf("failed to walk %s: %w", path, err)
-		}
-	}
-
-	return nil
-}
 
-func (g *GitReader) Read(ctx context.Context, files []*File) (n int, err error) {
-	idx := 0
+		// create a new scanner for reading the output
+		g.scanner = bufio.NewScanner(r)
+	}
 
 LOOP:
-	for idx < len(files) {
+
+	for n < len(files) {
```

Comment:
I having trouble understanding this. It looks like files is an array passed into this function, and we're overwriting it with the output of git ls-files. Things I am concerned about:

- what if `files` isn't large enough?
- what if `files` is larger than the number of results we get back from `ls-files`?

Separately, are we checking the exit code of the git subproccess? Ideally we'd error out if something went wrong there.

Reply:
Here's how we are executing `git ls-files`:

```go
if g.scanner == nil {
		// create a pipe to capture the command output
		r, w := io.Pipe()

		// create a command which will execute from the specified sub path within root
		cmd := exec.Command("git", "ls-files")
		cmd.Dir = filepath.Join(g.root, g.path)
		cmd.Stdout = w

		// execute the command in the background
		g.eg.Go(func() error {
			return w.CloseWithError(cmd.Run())
		})

		// create a new scanner for reading the output
		g.scanner = bufio.NewScanner(r)
	}
```

This does a few things:

- Creates an in memory pipe, with a reader (r) and write (w). 
- Sets the working directory for the command to `${g.root}/${g.path}`
- Sets stdout for the command to the write end of the pipe
- Runs the command in a separate go routine with `cmd.Run`, passing any error (exit code) returned into the close method of the write end of the pipe. This will emit the error on next read on the other end, nor `io.EOF` if the error was `nil`.
- We initialise a `bufio.Scanner` to read stdout line by line

The overall idea for the method is that you pass in an array with a length matching your maximum batch size. We read from the stdout of `git ls-files`, line by line, until we have filled that array, at which point we return to the user and indicate how many of them we filled `n`. This is borrowing from how other `Reader` interfaces work within go and is somewhat 'idiomatic'. 

It has the added benefit of re-using the array between subsequent calls, reducing GC.

Reply:
Oh this helps! This feels very much like c-style `read` semantics.

I was very much misreading this code. TIL about go's "named return values"